File: src/services/compliance/deadlines.ts
import { prisma } from "@/lib/db";
import type {
  DeadlineStatus,
  TriggerEventType,
  Severity,
  ComplianceDeadline,
} from "@prisma/client";
import { calculateDeadline, calculateDaysRemaining } from "./calculator";
import { calculateSeverity } from "./severity";

// ── Types ─────────────────────────────────────────────────────────────────

interface CreateDeadlineParams {
  projectId: string;
  clauseId: string;
  triggerEventType: TriggerEventType;
  triggerEventId?: string;
  triggerDescription: string;
  triggeredAt: Date;
  triggeredBy?: string;
}

interface DeadlineWithCountdown {
  id: string;
  projectId: string;
  clauseId: string;
  sectionRef: string | null;
  clauseTitle: string;
  clauseKind: string;
  triggerEventType: TriggerEventType;
  triggerDescription: string;
  triggeredAt: Date;
  calculatedDeadline: Date;
  daysRemaining: number;
  hoursRemaining: number;
  severity: Severity;
  status: DeadlineStatus;
  noticeId: string | null;
  noticeCreatedAt: Date | null;
  waivedAt: Date | null;
  waiverReason: string | null;
  createdAt: Date;
}

interface DeadlineListResult {
  activeCritical: number;
  activeWarning: number;
  totalActive: number;
  total: number;
  deadlines: DeadlineWithCountdown[];
}

// ── Create ────────────────────────────────────────────────────────────────

export async function createDeadline(
  params: CreateDeadlineParams
): Promise<ComplianceDeadline> {
  const {
    projectId,
    clauseId,
    triggerEventType,
    triggerEventId,
    triggerDescription,
    triggeredAt,
    triggeredBy,
  } = params;

  // Get the clause for deadline calculation
  const clause = await prisma.contractClause.findUnique({
    where: { id: clauseId },
  });
  if (!clause) {
    throw new Error(`Clause not found: ${clauseId}`);
  }
  if (!clause.deadlineDays || !clause.deadlineType) {
    throw new Error(
      `Clause ${clauseId} has no deadline configured (deadlineDays or deadlineType missing)`
    );
  }

  // Calculate the deadline date
  const result = await calculateDeadline({
    triggerDate: triggeredAt,
    deadlineDays: clause.deadlineDays,
    deadlineType: clause.deadlineType,
    curePeriodDays: clause.curePeriodDays ?? undefined,
    curePeriodType: clause.curePeriodType ?? undefined,
    projectId,
  });

  // Determine initial severity
  const { severity } = calculateSeverity(result.calculatedDeadline);

  // Create deadline record
  const deadline = await prisma.complianceDeadline.create({
    data: {
      projectId,
      clauseId,
      triggerEventType,
      triggerEventId: triggerEventId || null,
      triggerDescription,
      triggeredAt,
      triggeredBy: triggeredBy || null,
      calculatedDeadline: result.calculatedDeadline,
      status: "ACTIVE",
      severity,
    },
  });

  // Audit log
  await prisma.complianceAuditLog.create({
    data: {
      projectId,
      eventType: "DEADLINE_CREATED",
      entityType: "ComplianceDeadline",
      entityId: deadline.id,
      userId: triggeredBy || null,
      actorType: triggeredBy ? "USER" : "SYSTEM",
      action: "created",
      details: JSON.parse(
        JSON.stringify({
          sectionRef: clause.sectionRef,
          triggerEventType,
          calculatedDeadline: result.calculatedDeadline.toISOString(),
          calendarDaysCount: result.calendarDaysCount,
          severity,
        })
      ),
    },
  });

  return deadline;
}

// ── Query ─────────────────────────────────────────────────────────────────

export async function getProjectDeadlines(
  projectId: string,
  filters?: {
    status?: DeadlineStatus;
    severity?: Severity;
    sortBy?: "deadline" | "severity" | "created";
  }
): Promise<DeadlineListResult> {
  const deadlines = await prisma.complianceDeadline.findMany({
    where: {
      projectId,
      ...(filters?.status && { status: filters.status }),
      ...(filters?.severity && { severity: filters.severity }),
    },
    include: {
      clause: {
        select: {
          sectionRef: true,
          title: true,
          kind: true,
        },
      },
    },
    orderBy: { calculatedDeadline: "asc" },
  });

  const now = new Date();

  const withCountdown: DeadlineWithCountdown[] = deadlines.map((d) => {
    const daysRemaining = calculateDaysRemaining(d.calculatedDeadline, now);
    const hoursRemaining = Math.max(
      0,
      Math.ceil(
        (d.calculatedDeadline.getTime() - now.getTime()) / (1000 * 60 * 60)
      )
    );

    return {
      id: d.id,
      projectId: d.projectId,
      clauseId: d.clauseId,
      sectionRef: d.clause.sectionRef,
      clauseTitle: d.clause.title,
      clauseKind: d.clause.kind,
      triggerEventType: d.triggerEventType,
      triggerDescription: d.triggerDescription,
      triggeredAt: d.triggeredAt,
      calculatedDeadline: d.calculatedDeadline,
      daysRemaining,
      hoursRemaining,
      severity: d.severity,
      status: d.status,
      noticeId: d.noticeId,
      noticeCreatedAt: d.noticeCreatedAt,
      waivedAt: d.waivedAt,
      waiverReason: d.waiverReason,
      createdAt: d.createdAt,
    };
  });

  // Sort
  if (filters?.sortBy === "severity") {
    const order: Record<string, number> = {
      EXPIRED: 0,
      CRITICAL: 1,
      WARNING: 2,
      INFO: 3,
      LOW: 4,
    };
    withCountdown.sort(
      (a, b) => (order[a.severity] ?? 5) - (order[b.severity] ?? 5)
    );
  } else if (filters?.sortBy === "created") {
    withCountdown.sort(
      (a, b) => b.createdAt.getTime() - a.createdAt.getTime()
    );
  }
  // default sort is by calculatedDeadline (already from DB query)

  const active = withCountdown.filter((d) => d.status === "ACTIVE");

  return {
    activeCritical: active.filter(
      (d) => d.severity === "CRITICAL" || d.severity === "EXPIRED"
    ).length,
    activeWarning: active.filter((d) => d.severity === "WARNING").length,
    totalActive: active.length,
    total: withCountdown.length,
    deadlines: withCountdown,
  };
}

// ── Waive ─────────────────────────────────────────────────────────────────

export async function waiveDeadline(
  deadlineId: string,
  userId: string,
  reason: string
): Promise<ComplianceDeadline> {
  const deadline = await prisma.complianceDeadline.update({
    where: { id: deadlineId },
    data: {
      status: "WAIVED",
      waivedAt: new Date(),
      waivedBy: userId,
      waiverReason: reason,
    },
  });

  await prisma.complianceAuditLog.create({
    data: {
      projectId: deadline.projectId,
      eventType: "DEADLINE_WAIVED",
      entityType: "ComplianceDeadline",
      entityId: deadlineId,
      userId,
      actorType: "USER",
      action: "waived",
      details: JSON.parse(JSON.stringify({ reason })),
    },
  });

  return deadline;
}

// ── Severity refresh (called by cron) ─────────────────────────────────────

export async function updateAllDeadlineSeverities(): Promise<{
  updated: number;
  expired: number;
}> {
  const activeDeadlines = await prisma.complianceDeadline.findMany({
    where: { status: "ACTIVE" },
  });

  let updated = 0;
  let expired = 0;

  for (const deadline of activeDeadlines) {
    const { severity } = calculateSeverity(deadline.calculatedDeadline);

    if (severity !== deadline.severity) {
      const newStatus: DeadlineStatus =
        severity === "EXPIRED" ? "EXPIRED" : deadline.status;

      await prisma.complianceDeadline.update({
        where: { id: deadline.id },
        data: { severity, status: newStatus },
      });
      updated++;

      if (severity === "EXPIRED") {
        expired++;

        await prisma.complianceAuditLog.create({
          data: {
            projectId: deadline.projectId,
            eventType: "DEADLINE_EXPIRED",
            entityType: "ComplianceDeadline",
            entityId: deadline.id,
            actorType: "SYSTEM",
            action: "expired",
            details: JSON.parse(
              JSON.stringify({
                calculatedDeadline:
                  deadline.calculatedDeadline.toISOString(),
                consequence: "Claim forfeiture risk",
              })
            ),
          },
        });
      }
    }
  }

  return { updated, expired };
}

File: backend/app/services/compliance/deadlines.py
"""Deadline management service.

Handles creation, severity updates, waiver, and lifecycle of compliance deadlines.
"""

import logging
from datetime import datetime

from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.compliance import (
    ComplianceAuditLog,
    ComplianceDeadline,
    ContractClause,
)
from app.models.enums import DeadlineStatus, DeadlineType, Severity, TriggerEventType

from .calculator import calculate_deadline
from .severity import classify_severity, severity_escalated

logger = logging.getLogger(__name__)


async def create_deadline(
    db: AsyncSession,
    project_id: str,
    clause_id: str,
    trigger_event_type: TriggerEventType,
    trigger_description: str,
    triggered_at: datetime,
    trigger_event_id: str | None = None,
    triggered_by: str | None = None,
) -> ComplianceDeadline | None:
    """Create a new compliance deadline from a trigger event.

    Looks up the clause's deadline parameters and calculates the deadline date.
    """
    # Load clause
    result = await db.execute(
        select(ContractClause).where(
            ContractClause.id == clause_id,
            ContractClause.project_id == project_id,
        )
    )
    clause = result.scalar_one_or_none()
    if not clause:
        logger.warning("Clause %s not found", clause_id)
        return None

    if not clause.deadline_days or not clause.deadline_type:
        logger.warning("Clause %s has no deadline parameters", clause_id)
        return None

    # Calculate deadline date
    calc = await calculate_deadline(
        db,
        project_id,
        trigger_date=triggered_at,
        deadline_days=clause.deadline_days,
        deadline_type=clause.deadline_type,
        cure_period_days=clause.cure_period_days,
        cure_period_type=clause.cure_period_type,
    )

    deadline = ComplianceDeadline(
        project_id=project_id,
        clause_id=clause_id,
        trigger_event_type=trigger_event_type,
        trigger_event_id=trigger_event_id,
        trigger_description=trigger_description,
        triggered_at=triggered_at,
        triggered_by=triggered_by,
        calculated_deadline=calc["calculatedDeadline"],
        status=DeadlineStatus.ACTIVE,
        severity=calc["severity"],
    )
    db.add(deadline)
    await db.flush()
    await db.refresh(deadline)

    # Audit log
    audit = ComplianceAuditLog(
        project_id=project_id,
        event_type="DEADLINE_CREATED",
        entity_type="ComplianceDeadline",
        entity_id=deadline.id,
        user_id=triggered_by,
        actor_type="SYSTEM",
        action="create_deadline",
        details={
            "clauseId": clause_id,
            "clauseTitle": clause.title,
            "triggerType": trigger_event_type.value,
            "triggerDescription": trigger_description,
            "calculatedDeadline": calc["calculatedDeadline"].isoformat(),
            "severity": calc["severity"].value,
        },
    )
    db.add(audit)

    await db.flush()
    logger.info(
        "Created deadline %s for clause %s (due %s, severity %s)",
        deadline.id, clause_id,
        calc["calculatedDeadline"].isoformat(),
        calc["severity"].value,
    )
    return deadline


async def get_deadlines(
    db: AsyncSession,
    project_id: str,
    status: DeadlineStatus | None = None,
    severity: Severity | None = None,
) -> list[ComplianceDeadline]:
    """List deadlines for a project with optional filters."""
    query = (
        select(ComplianceDeadline)
        .where(ComplianceDeadline.project_id == project_id)
        .order_by(ComplianceDeadline.calculated_deadline.asc())
    )

    if status:
        query = query.where(ComplianceDeadline.status == status)
    if severity:
        query = query.where(ComplianceDeadline.severity == severity)

    result = await db.execute(query)
    return list(result.scalars().all())


async def get_deadline_by_id(
    db: AsyncSession,
    deadline_id: str,
    project_id: str,
) -> ComplianceDeadline | None:
    """Get a single deadline."""
    result = await db.execute(
        select(ComplianceDeadline).where(
            ComplianceDeadline.id == deadline_id,
            ComplianceDeadline.project_id == project_id,
        )
    )
    return result.scalar_one_or_none()


async def update_deadline_status(
    db: AsyncSession,
    deadline_id: str,
    project_id: str,
    new_status: DeadlineStatus,
    user_id: str | None = None,
    notice_id: str | None = None,
) -> ComplianceDeadline | None:
    """Update a deadline's status."""
    deadline = await get_deadline_by_id(db, deadline_id, project_id)
    if not deadline:
        return None

    old_status = deadline.status
    deadline.status = new_status

    if new_status == DeadlineStatus.NOTICE_DRAFTED and notice_id:
        deadline.notice_id = notice_id
        deadline.notice_created_at = datetime.utcnow()
    elif new_status == DeadlineStatus.NOTICE_SENT:
        pass  # sentAt tracked on the notice itself

    # Audit
    audit = ComplianceAuditLog(
        project_id=project_id,
        event_type="DEADLINE_STATUS_CHANGE",
        entity_type="ComplianceDeadline",
        entity_id=deadline_id,
        user_id=user_id,
        actor_type="USER" if user_id else "SYSTEM",
        action="update_status",
        details={
            "oldStatus": old_status.value,
            "newStatus": new_status.value,
            "noticeId": notice_id,
        },
    )
    db.add(audit)

    await db.flush()
    return deadline


async def waive_deadline(
    db: AsyncSession,
    deadline_id: str,
    project_id: str,
    user_id: str,
    reason: str,
) -> ComplianceDeadline | None:
    """Waive a deadline with reason tracking."""
    deadline = await get_deadline_by_id(db, deadline_id, project_id)
    if not deadline:
        return None

    deadline.status = DeadlineStatus.WAIVED
    deadline.waived_at = datetime.utcnow()
    deadline.waived_by = user_id
    deadline.waiver_reason = reason
    deadline.severity = Severity.LOW

    # Audit
    audit = ComplianceAuditLog(
        project_id=project_id,
        event_type="DEADLINE_WAIVED",
        entity_type="ComplianceDeadline",
        entity_id=deadline_id,
        user_id=user_id,
        actor_type="USER",
        action="waive_deadline",
        details={"reason": reason},
    )
    db.add(audit)

    await db.flush()
    return deadline


async def recalculate_severities(
    db: AsyncSession,
    project_id: str,
) -> dict:
    """Recalculate severity for all active deadlines in a project.

    Returns counts of changes by severity level.
    """
    now = datetime.utcnow()
    active_statuses = [DeadlineStatus.ACTIVE, DeadlineStatus.NOTICE_DRAFTED]

    result = await db.execute(
        select(ComplianceDeadline).where(
            ComplianceDeadline.project_id == project_id,
            ComplianceDeadline.status.in_(active_statuses),
        )
    )
    deadlines = result.scalars().all()

    changes = {"escalated": 0, "changed": 0, "expired": 0, "total": len(deadlines)}

    for deadline in deadlines:
        old_severity = deadline.severity
        new_severity = classify_severity(
            deadline.calculated_deadline, now, deadline.status
        )

        if old_severity != new_severity:
            deadline.severity = new_severity
            changes["changed"] += 1

            if severity_escalated(old_severity, new_severity):
                changes["escalated"] += 1

            if new_severity == Severity.EXPIRED:
                changes["expired"] += 1
                deadline.status = DeadlineStatus.EXPIRED

    if changes["changed"] > 0:
        await db.flush()
        logger.info(
            "Recalculated severities for project %s: %d changes, %d escalated, %d expired",
            project_id, changes["changed"], changes["escalated"], changes["expired"],
        )

    return changes
